basis/BasisFunction.py

```diff
@@ -0,0 +1,110 @@
+from abc import ABC, abstractmethod
+
+class BasisFunction(ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ''max_exactness'' represents the maximum level of exactness, or 
+    accuracy, of a surrogate function this class can provide enough 
+    information for. This class can provide enough information for 
+    constructing a surrogate function of any accuracy at or below 
+    max_exactness.
+    ''extrema_per_level_num'' property represents the number of unique 
+    extrema that correspond to increasing levels of exactness. It is 
+    assumed that for an exactness of 0 the number of unique extrema is 1.
+    ''extrema_list'' property represents all the unique extrema of the 
+    basis function in order of computation. Each extrema is associated 
+    with a Smolyak index described by the class IndexGrid, and the 
+    Smolyak index of a given extrema is equal to the extrema's position
+    in the array extrema_list. It is assumed the first extrema of any
+    basis function is 0.
+
+    Parameters
+    ----------
+    max_exactness : int
+        level of exactness class will describe
+
+    """
+
+    def __init__(self,max_exactness):
+        """Initializer"""
+        self._extrema_list = [0]
+        self._extrema_per_level_num = [1]
+        self._max_exactness = max_exactness
+        self.update_extrema_index_per_level()
+
+    @property
+    def max_exactness(self):
+        """Maximum level of exactness described"""
+        return self._max_exactness
+
+    @property
+    def extrema_list(self):
```

Comment:
Property names should not include the data type, that is an implementation detail:
```suggestion
    def extrema(self):
```